cannon/mipsevm/tests/fuzz_evm_common_test.go

```diff
@@ -54,367 +52,440 @@ func FuzzStateSyscallMmap(f *testing.F) {
+	}
+
+	diffTester := NewDiffTester(NoopTestNamer[testCase]).
+		InitState(initState, mtutil.WithPCAndNextPC(0), mtutil.WithPreimageKey(preimageKey), mtutil.WithPreimageOffset(128)).
+		SetExpectations(setExpectations).
+		PostCheck(postCheck)
+
 	f.Fuzz(func(t *testing.T, addr arch.Word, count arch.Word, seed int64) {
-		for _, v := range versions {
-			t.Run(v.Name, func(t *testing.T) {
-				// Make sure pc does not overlap with preimage data in memory
-				pc := Word(0)
-				if addr <= 8 {
-					addr += 8
-				}
-				effAddr := addr & arch.AddressMask
-				preexistingMemoryVal := [8]byte{0x12, 0x34, 0x56, 0x78, 0x87, 0x65, 0x43, 0x21}
-				preimageData := []byte("hello world")
-				preimageKey := preimage.Keccak256Key(crypto.Keccak256Hash(preimageData)).PreimageKey()
-				oracle := testutil.StaticOracle(t, preimageData)
-
-				goVm := v.VMFactory(oracle, os.Stdout, os.Stderr, testutil.CreateLogger(),
-					mtutil.WithRandomization(seed), mtutil.WithPreimageKey(preimageKey), mtutil.WithPreimageOffset(128), mtutil.WithPCAndNextPC(pc))
-				state := goVm.GetState()
-				state.GetRegistersRef()[2] = arch.SysWrite
-				state.GetRegistersRef()[4] = exec.FdPreimageWrite
-				state.GetRegistersRef()[5] = addr
-				state.GetRegistersRef()[6] = count
-				testutil.StoreInstruction(state.GetMemory(), state.GetPC(), syscallInsn)
-				state.GetMemory().SetWord(effAddr, arch.ByteOrderWord.Word(preexistingMemoryVal[:]))
-				step := state.GetStep()
-
-				expectBytesWritten := count
-				alignment := addr & arch.ExtMask
-				sz := arch.WordSizeBytes - alignment
-				if sz < expectBytesWritten {
-					expectBytesWritten = sz
-				}
-
-				expected := mtutil.NewExpectedState(t, state)
-				expected.ExpectStep()
-				expected.PreimageOffset = 0
-				expected.ActiveThread().Registers[2] = expectBytesWritten
-				expected.ActiveThread().Registers[7] = 0 // No error
-				expected.PreimageKey = preimageKey
-				if expectBytesWritten > 0 {
-					// Copy original preimage key, but shift it left by expectBytesWritten
-					copy(expected.PreimageKey[:], preimageKey[expectBytesWritten:])
-					// Copy memory data to rightmost expectedBytesWritten
-					copy(expected.PreimageKey[32-expectBytesWritten:], preexistingMemoryVal[alignment:])
-				}
-
-				stepWitness, err := goVm.Step(true)
-				require.NoError(t, err)
-				require.False(t, stepWitness.HasPreimage())
-
-				expected.Validate(t, state)
-				testutil.ValidateEVM(t, stepWitness, step, goVm, v.StateHashFn, v.Contracts)
-			})
+		if addr <= 8 {
+			addr += 8
+		}
+
+		po := func() mipsevm.PreimageOracle {
+			return testutil.StaticOracle(t, preimageData)
 		}
+
+		tests := []testCase{{addr, count}}
+		diffTester.Run(t, tests, fuzzTestOptions(vms, seed, WithPreimageOracle(po))...)
 	})
 }
+
+func fuzzTestOptions(vms []VersionedVMTestCase, seed int64, opts ...TestOption) []TestOption {
+	testOpts := []TestOption{
+		WithVms(vms),
+		WithRandomSeed(seed),
+		SkipAutomaticMemoryReservationTests(),
```

Comment:
Set this up to skip the automatic tests initially, but we can add this back in if we want.